### PR TITLE
additional changes for lower memory usage

### PR DESCRIPTION
**Pull request by Thomas-MMJ**
lower batch size with higher gradient accumulation uses less memory for same training benefit; use_8bit_adam greatly reduces memory for adam; gradient_checkpointing greatly reduces memory; mixed_precision bf16 is faster for same memory usage.

Note that I can't test these locally yet (on windows) so not positive they are all of benefit.  Will probably test tomorrow on colab.

### CODE REVIEW

**Comment by 1blackbar** [comment]
Theres a cell missing to download trained bin file , also cell missing to prune it, for newbies this is kinda not useful at all if you cant downoad model, theres 3 bin files , takes ages to download one just to figure out its not the right one....
